library/scrollphathd/api/api.py

```diff
@@ -0,0 +1,73 @@
+from sys import exit
+import scrollphathd
+from scrollphathd.fonts import font3x5
+from queue import Queue
+from threading import Thread
+from .action import Action
+from http import HTTPStatus
+
+try:
+    from flask import Blueprint, render_template, abort, request
+except ImportError:
+    exit("flask must be installed in order to use the api. Install with pip install flask")
+
+scrollphathd_blueprint = Blueprint('scrollhat', __name__)
+api_queue = Queue()
+
+@scrollphathd_blueprint.route('/scroll', methods=["POST"])
+def scroll():
+    data = request.get_json()
+    try:
+        api_queue.put(Action("scroll", (data["x"], data["y"])))
+    except KeyError:
+        return {"error": "keys x and y not posted."}, HTTPStatus.UNPROCESSABLE_ENTITY.value
+    else:
+        return HTTPStatus.OK.value
+
+
+@scrollphathd_blueprint.route('/show', methods=["POST"])
+def show(text):
+    data = request.get_json()
+    try:
+        api_queue.put(Action("write", data["text"]))
+    except KeyError:
+        return {"error": "key 'text' not set"}, HTTPStatus.UNPROCESSABLE_ENTITY.value
+    else:
+        return HTTPStatus.OK.value
+
+@scrollphathd_blueprint.route('/clear', methods=["POST"])
+def clear():
+    api_queue.put(Action("clear", {}))
+    return HTTPStatus.OK.value
+
+@scrollphathd.route('/flip', methods=["POST"])
+def flip():
+    data = request.get_json()
+    try:
+        api_queue.put(Action("flip", (bool(data["x"]), bool(data["y"]))))
+    except TypeError:
+        return {"error": "Could not cast data correctly. Both `x` and `y` must be set to true or false."}, \
+               HTTPStatus.UNPROCESSABLE_ENTITY.value
+    else:
+        return HTTPStatus.OK.value
+
+def run():
+    while True:
+        action = api_queue.get(block=True)
+        if action.action_type == "write":
+            scrollphathd.write_string(action.data, font=font3x5)
+            scrollphathd.show()
+
+        if action.action_type == "clear":
+            scrollphathd.clear()
+
+        if action.action_type == "scroll":
+            scrollphathd.scroll(action.data[0], action.data[1])
+            scrollphathd.show()
+
+        if action.action_type == "flip":
+            scrollphathd.flip(x=action.data[0], y=action.data[1])
+
+def start_background_thread():
+    api_thread = Thread(target=run)
```

Comment:
I'd use a boilerplate stoppable-thread pattern here, and register an exit handler upon `api_thread.start()` to make sure it's killed cleanly when the script exits. Honestly, I don't know what the best pattern is, but I've used things like:

```python
class StoppableThread(threading.Thread):
    """Basic Stoppable Thread Wrapper
    Adds event for stopping the execution
    loop and exiting cleanly."""
    def __init__(self):
        threading.Thread.__init__(self)
        self.stop_event = threading.Event()
        self.daemon = True

    def start(self):
        if not self.isAlive():
            self.stop_event.clear()
            threading.Thread.start(self)

    def stop(self):
        if self.isAlive():
            self.stop_event.set()
            self.join()
```

Reply:
hmmm I nicked your boilerplate code but I'm not sure that will work (I know it was an example) given stop isn't taken from the parent class, so it won't ever get called by the usual thread pattern. Will look into it after work

Reply:
also why does it need to be a daemon thread?

Reply:
Good question- possibly because I didn't understand Python threading comprehensively enough to make the right choice when I adopted that pattern. Daemon threads have the nice side-effect of bailing with your program, so if you're like me and forget to stop and join then gracefully all the pain of a hanging exit gets swept under the rug.

I still don't comprehensively understand threading, but I suspect it's unnecessary to daemonize the thread as long as it's exited cleanly in all cases. I don't know what implications this has for program exits in case of an unhandled exception, etc, though. This is probably why I prefer to damonize and *try* to explicitly stop the thread cleanly on exit (using `atexit`) but let it bail of its own accord in other cases.

Reply:
see I thought it was the other way round?

Reply:
yeah, I just googled this "A daemon thread is a thread that does not prevent the JVM from exiting when the program finishes but the thread is still running"...